### PR TITLE
fix: make table command work with any set of event definitions, always write header (even for empty files)

### PR DESCRIPTION
**Pull request by dlaehnemann**
It's all in the title. While at it, I also updated some of the dependencies. Any recommendations on the code welcome, it's been a while since I have done anything in Rust (and it has been a battle with the compiler;).

### CODE REVIEW

File: src/graph/table.rs [hunk 1]
```diff
@@ -32,15 +33,26 @@ pub(crate) struct TableArgs {
     #[arg()]
     records: PathBuf,
 
-    /// Path to breakend records in BCF format
+    /// Path to reference
```
Comment (tedil) [inline src/graph/table.rs]: ... whoops, thanks for fixing.

File: src/graph/table.rs [hunk 2]
```diff
@@ -204,15 +233,15 @@ fn tables_from_annotated_graph(
                     gene_names,
                     regulatory_features,
                     repeats,
-                    prob_present: varlociraptor_info.prob_present,
+                    prob_joint_event: varlociraptor_info.event_probs.values().sum(),
                     prob_absent: varlociraptor_info.prob_absent,
-                    prob_artifact: varlociraptor_info.prob_artifact,
+                    prob_artifact: varlociraptor_info.prob_absent,
```
Comment (tedil) [inline src/graph/table.rs]:
```suggestion
                    prob_artifact: varlociraptor_info.prob_artifact,
```

File: src/graph/table.rs [hunk 3]
```diff
@@ -103,52 +117,67 @@ pub(crate) fn group_event_records(
...
+    record.info()
+        .get(&key)
+        .map(|f| match f.value() {
+            Some(Value::Float(f)) => *f,
+            _ => panic!("Expected float value for {}", prob_name),
+        })
+        .unwrap_or_else(|| panic!("'{}' info field not found", prob_name))
+}
+
+fn varlociraptor_info(records: &[Record], prob_info_fields: &Vec<String>) -> VarlociraptorInfo {
```
Comment (tedil) [inline src/graph/table.rs]:
```suggestion
fn varlociraptor_info(records: &[Record], prob_info_fields: &[String]) -> VarlociraptorInfo {
```

File: src/graph/table.rs [hunk 4]
```diff
@@ -103,52 +117,67 @@ pub(crate) fn group_event_records(
...
     prob_artifact: f32,
+    event_probs: HashMap<String, f32>,
     af_nanopore: Option<f32>,
 }
 
 fn tables_from_annotated_graph(
     graph: &AnnotatedGraph,
     event_records: &HashMap<String, Vec<Record>>,
     reference: &Reference,
+    prob_info_fields: &Vec<String>,
```
Comment (tedil) [inline src/graph/table.rs]:
```suggestion
    prob_info_fields: &[String],
```

File: src/graph/table.rs [hunk 5]
```diff
@@ -103,52 +117,67 @@ pub(crate) fn group_event_records(
...
 }
 
-#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
+#[derive(Serialize, Deserialize, Debug, Clone)]
 pub(crate) struct VarlociraptorInfo {
-    prob_present: f32,
     prob_absent: f32,
     prob_artifact: f32,
+    event_probs: HashMap<String, f32>,
     af_nanopore: Option<f32>,
```
Comment (tedil) [inline src/graph/table.rs:173]: This is hardcoded, but basically defined by the scenario / sample configuration, perhaps we can determine this dynamically as well?

**Review by tedil — CHANGES_REQUESTED** [review]
Nice work, thank you very much!
I only have some minor remarks / suggestions, otherwise looks good to me :)

File: src/graph/table.rs (hunk 1 above)
Comment (dlaehnemann) [inline src/graph/table.rs]: We could even make this more descriptive:
```suggestion
    /// Path to reference file in fasta format
```
And does `bio::io::fasta::Reader()` handle gzipped files? From a quick look, I don't see any functionality for this in the code, but maybe this is some levels further down (e.g. in `std::fs`)?

File: src/graph/table.rs (hunk 2 above)
Comment (dlaehnemann) [reply src/graph/table.rs]: cooopy-pasta

:clap: :clap:

:clap: :clap: :clap: :clap:

File: src/graph/table.rs (hunk 5 above)
Comment (dlaehnemann) [reply src/graph/table.rs:173]: Well, in the snakemake workflow using this, this is determined by a mixture of the `scenario.yaml` and the `config.yaml`. The scenario determines which events exist and CAN be used. And the `config.yaml` entries under `filter: fdr-control: events: <joint_event_name>: []` determine which events ARE actually used in the `joint_event`. These don't have to be all the available events. Does this explanation make it clearer? Should I make the command-line argument explanations clearer to get this across?

**Comment by dlaehnemann** [comment]
I applied all the suggestions and tested them locally. Seems to compile, and seems to properly create the testing tsv files I have been generating to test the new version. So feel free to have another look.

File: src/graph/table.rs (hunk 5 above)
Comment (tedil) [reply src/graph/table.rs:173]: Sorry, the highlighting is a bit off here; my comment was with respect to the af_nanopore entry

Comment (dlaehnemann) [reply src/graph/table.rs:173]: Ah, I see. But how would we want to report this, then? There can be any number of samples in the scenario and we can determine this dynamically. But how do we output this in the table and how do we process this in the workflow? Use some prefix and then put all the allele frequencies in the resulting report (like you do with the `prob_` prefix there)? One problem would be serializing the records though. I was thinking about this for the (sub-)event probs, and then decided to simply sum up their probs into one value to get around it.

Comment (tedil) [reply src/graph/table.rs:173]: Yeah, it's not that straightforward, so let's postpone this to the future™

**Pull request merged**